File: examples/parsefacts/Python/parsefacts.py

from facts_pb2 import TDocument
import sys

def ReadVarint32(f):
  offset = 0
  result = 0;
  byte = f.read(1)
  while byte != "":
    if 0 == offset:
      result = ord(byte) & 0x7F
    else:
      result = result | ( ( ord(byte) & 0x7F ) << offset )

    offset += 7

    if (ord(byte) & 0x80) == 0 :
      break

    byte = f.read(1)

  return result


for a in sys.argv[1:]:
  ff = open(a, "rb")
  

  size = ReadVarint32(ff)
  while (size > 0):
    data = ff.read(size)
    doc = TDocument()
    doc.ParseFromString(data)

    print "id =", doc.Id
    print "url =", doc.Url
  
    for fg in doc.FactGroup:
      print "fact type =", fg.Type

      for f in fg.Fact:
        print "index =", f.Index

        for field in f.Field:
          print field.Name, "=", field.Value

    size = ReadVarint32(ff)


  ff.close;

File: examples/parsefacts/Ruby/parsefacts.rb
#!/usr/bin/env ruby

# Please make sure that the `beefcake` gem is installed and
# the `generate.sh` script has been executed.

require 'beefcake'
require_relative 'facts.pb'

def read_varint32(f)
  offset, result = 0, 0

  until f.eof? || (byte = f.readbyte).zero?
    result = if offset.zero?
      byte & 0x7F
    else
      result | ((byte & 0x7F) << offset)
    end

    offset += 7
    break if (byte & 0x80).zero?
  end

  result
end

ARGV.each do |path|
  File.open(path) do |f|
    until (size = read_varint32(f)).zero?
      data = f.read(size)
      document = TDocument.decode(data)

      puts 'id = %d' % document.Id
      puts 'url = "%s"' % document.Url

      document.FactGroup.each do |fg|
        puts 'fact type = "%s"' % fg.Type

        fg.Fact.each do |fact|
          puts 'index = %d' % fact.Index

          fact.Field.each do |field|
            puts '%s = "%s"' % [field.Name, field.Value]
          end
        end
      end
    end
  end
end
